synapse/handlers/federation_event.py

```diff
@@ -1023,9 +1023,15 @@ async def _handle_marker_event(self, origin: str, marker_event: EventBase):
             return
```

Comment:
After merging `develop` (725cf222), one of the Complement tests started failing. I think it's related to some change in https://github.com/matrix-org/synapse/pull/10692

```sh
$ COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS=1 COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh TestBackfillingHistory/parallel/Historical_messages_are_visible_when_joining_on_federated_server_-_pre-made_insertion_event

=== RUN   TestBackfillingHistory
    msc2716_test.go:62: Deploy times: 6.225696908s blueprints, 4.390826875s containers
    client.go:370: POST hs1/_matrix/client/r0/register => 200 OK (24.494359ms)
=== RUN   TestBackfillingHistory/parallel
=== RUN   TestBackfillingHistory/parallel/Historical_messages_are_visible_when_joining_on_federated_server_-_pre-made_insertion_event
=== PAUSE TestBackfillingHistory/parallel/Historical_messages_are_visible_when_joining_on_federated_server_-_pre-made_insertion_event
=== CONT  TestBackfillingHistory/parallel/Historical_messages_are_visible_when_joining_on_federated_server_-_pre-made_insertion_event
=== CONT  TestBackfillingHistory
    client.go:370: POST hs1/_matrix/client/r0/createRoom => 200 OK (204.314404ms)
    client.go:370: POST hs1/_matrix/client/r0/join/!pFzlAcdlcfUYBlwrgD:hs1 => 200 OK (38.364663ms)
    client.go:370: PUT hs1/_matrix/client/r0/rooms/!pFzlAcdlcfUYBlwrgD:hs1/send/m.room.message/1 => 200 OK (37.496322ms)
=== CONT  TestBackfillingHistory/parallel/Historical_messages_are_visible_when_joining_on_federated_server_-_pre-made_insertion_event
    msc2716_test.go:979: SendEventSynced waiting for event ID $Mq1B-RVgu7d8QHY9cqZ4nnV5MLCHYadRVBPluTGfSXo
=== CONT  TestBackfillingHistory
    client.go:370: GET hs1/_matrix/client/r0/sync => 200 OK (40.755036ms)
    client.go:370: PUT hs1/_matrix/client/r0/rooms/!pFzlAcdlcfUYBlwrgD:hs1/send/org.matrix.msc2716.insertion/txn-m123 => 200 OK (14.01282ms)
    client.go:370: GET hs1/_matrix/client/r0/sync => 200 OK (19.589308ms)
    client.go:370: PUT hs1/_matrix/client/r0/rooms/!pFzlAcdlcfUYBlwrgD:hs1/send/m.room.message/2 => 200 OK (14.515922ms)
=== CONT  TestBackfillingHistory/parallel/Historical_messages_are_visible_when_joining_on_federated_server_-_pre-made_insertion_event
    msc2716_test.go:979: SendEventSynced waiting for event ID $SrhE54fDdLcQI6dyxFE-QaTfSZhGEB0TkjUagSQ5AKc
=== CONT  TestBackfillingHistory
    client.go:370: GET hs1/_matrix/client/r0/sync => 200 OK (19.774538ms)
    client.go:370: PUT hs1/_matrix/client/r0/rooms/!pFzlAcdlcfUYBlwrgD:hs1/send/m.room.message/3 => 200 OK (15.798959ms)
=== CONT  TestBackfillingHistory/parallel/Historical_messages_are_visible_when_joining_on_federated_server_-_pre-made_insertion_event
    msc2716_test.go:979: SendEventSynced waiting for event ID $CSvwLw2nHOW0O3MkI2EoTOTPqNFxI5Zdn73tzdKayBs
=== CONT  TestBackfillingHistory
    client.go:370: GET hs1/_matrix/client/r0/sync => 200 OK (18.961067ms)
    client.go:370: PUT hs1/_matrix/client/r0/rooms/!pFzlAcdlcfUYBlwrgD:hs1/send/m.room.message/4 => 200 OK (13.47559ms)
=== CONT  TestBackfillingHistory/parallel/Historical_messages_are_visible_when_joining_on_federated_server_-_pre-made_insertion_event
    msc2716_test.go:979: SendEventSynced waiting for event ID $VtcB-yfsAMsd34RofKyrV9vqKLHRmuQOIlBDU3x1IDA
=== CONT  TestBackfillingHistory
    client.go:370: GET hs1/_matrix/client/r0/sync => 200 OK (18.752454ms)
    client.go:370: POST hs1/_matrix/client/unstable/org.matrix.msc2716/rooms/!pFzlAcdlcfUYBlwrgD:hs1/batch_send => 200 OK (66.974695ms)
    client.go:370: POST hs2/_matrix/client/r0/join/!pFzlAcdlcfUYBlwrgD:hs1 => 200 OK (817.637887ms)
    client.go:370: GET hs2/_matrix/client/r0/rooms/!pFzlAcdlcfUYBlwrgD:hs1/messages => 200 OK (280.548957ms)
    client.go:370: GET hs2/_matrix/client/r0/rooms/!pFzlAcdlcfUYBlwrgD:hs1/messages => 200 OK (35.166542ms)
=== CONT  TestBackfillingHistory/parallel/Historical_messages_are_visible_when_joining_on_federated_server_-_pre-made_insertion_event
    msc2716_test.go:546: MatchResponse JSONCheckOff: did not see items: [$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY] - http://localhost:55288/_matrix/client/r0/rooms/%21pFzlAcdlcfUYBlwrgD:hs1/messages?dir=b&limit=100 => {"chunk":[{"type":"m.room.member","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@charlie:hs2","content":{"membership":"join","displayname":"charlie","avatar_url":null},"state_key":"@charlie:hs2","origin_server_ts":1630527561142,"unsigned":{"age":415},"event_id":"$sy89W9Z8SIB8TahLGggY77h_FJE1LRlY47nt7ot36DM","user_id":"@charlie:hs2","age":415},{"type":"m.room.message","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@alice:hs1","content":{"body":"Message 2","msgtype":"m.text"},"origin_server_ts":1630527560337,"unsigned":{"age":977},"event_id":"$VtcB-yfsAMsd34RofKyrV9vqKLHRmuQOIlBDU3x1IDA","user_id":"@alice:hs1","age":977},{"type":"m.room.message","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@alice:hs1","content":{"body":"Message 1","msgtype":"m.text"},"origin_server_ts":1630527560303,"unsigned":{"age":1011},"event_id":"$CSvwLw2nHOW0O3MkI2EoTOTPqNFxI5Zdn73tzdKayBs","user_id":"@alice:hs1","age":1011},{"type":"m.room.message","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@alice:hs1","content":{"body":"Message 0","msgtype":"m.text"},"origin_server_ts":1630527560267,"unsigned":{"age":1047},"event_id":"$SrhE54fDdLcQI6dyxFE-QaTfSZhGEB0TkjUagSQ5AKc","user_id":"@alice:hs1","age":1047},{"type":"m.room.message","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@maria:hs1","content":{"body":"Historical 0 (chunk=0)","msgtype":"m.text","org.matrix.msc2716.historical":true},"origin_server_ts":1630527560217,"unsigned":{"age":919},"event_id":"$YgVF4lkpq3ghIZd3C0evfHSJT1mhUA3cq92-triNh6E","user_id":"@maria:hs1","age":919},{"type":"m.room.message","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@maria:hs1","content":{"body":"Historical 1 (chunk=0)","msgtype":"m.text","org.matrix.msc2716.historical":true},"origin_server_ts":1630527560218,"unsigned":{"age":915},"event_id":"$lonJkljNkygvTMCeOqLpS4YLwrKch1cyPLirnrjhmqw","user_id":"@maria:hs1","age":915},{"type":"org.matrix.msc2716.insertion","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@the-bridge-user:hs1","content":{"org.matrix.msc2716.historical":true,"org.matrix.msc2716.next_chunk_id":"mynextchunkid123"},"origin_server_ts":1630527560234,"unsigned":{"age":1080},"event_id":"$kqz1sVD1G5kyK6tZ5_UR6uLNb6dKZihbXHYi1yScmDc","user_id":"@the-bridge-user:hs1","age":1080},{"type":"org.matrix.msc2716.chunk","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@the-bridge-user:hs1","content":{"org.matrix.msc2716.chunk_id":"mynextchunkid123","org.matrix.msc2716.historical":true},"origin_server_ts":1630527560218,"unsigned":{"age":911},"event_id":"$SyQzaYC7H4vlGUrQ0YdmccOBkCg71sL0KnMn425oj1U","user_id":"@the-bridge-user:hs1","age":911},{"type":"m.room.message","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@alice:hs1","content":{"body":"Message 0","msgtype":"m.text"},"origin_server_ts":1630527560173,"unsigned":{"age":1141},"event_id":"$Mq1B-RVgu7d8QHY9cqZ4nnV5MLCHYadRVBPluTGfSXo","user_id":"@alice:hs1","age":1141},{"type":"m.room.member","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@alice:hs1","content":{"membership":"join","displayname":"alice"},"state_key":"@alice:hs1","origin_server_ts":1630527560136,"unsigned":{"age":1045},"event_id":"$InXEtqdkYCAkjYvcYb_JctayzOP_1l_JIUvMSuwoHVs","user_id":"@alice:hs1","age":1045},{"type":"m.room.name","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@the-bridge-user:hs1","content":{"name":"the hangout spot"},"state_key":"","origin_server_ts":1630527560093,"unsigned":{"age":1088},"event_id":"$-OR4lb5yd1mjKROL4_fQXcp5tdVPdepJ0a0Jh0_xilU","user_id":"@the-bridge-user:hs1","age":1088},{"type":"m.room.history_visibility","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@the-bridge-user:hs1","content":{"history_visibility":"shared"},"state_key":"","origin_server_ts":1630527560063,"unsigned":{"age":1118},"event_id":"$d6nTx76gH2FrLTYfuWDGmTORvUaFciDanehIqWWLZYU","user_id":"@the-bridge-user:hs1","age":1118},{"type":"m.room.join_rules","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@the-bridge-user:hs1","content":{"join_rule":"public"},"state_key":"","origin_server_ts":1630527560039,"unsigned":{"age":1142},"event_id":"$QXoHXXLkWxVAlJ_BCiQwQnd3ypt9498g6oBizto5Bnk","user_id":"@the-bridge-user:hs1","age":1142},{"type":"m.room.power_levels","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@the-bridge-user:hs1","content":{"users":{"@the-bridge-user:hs1":100},"users_default":0,"events":{"m.room.name":50,"m.room.power_levels":100,"m.room.history_visibility":100,"m.room.canonical_alias":50,"m.room.avatar":50,"m.room.tombstone":100,"m.room.server_acl":100,"m.room.encryption":100},"events_default":0,"state_default":50,"ban":50,"kick":50,"redact":50,"invite":50,"historical":100},"state_key":"","origin_server_ts":1630527560009,"unsigned":{"age":1172},"event_id":"$bE3VMr1GJkI5BtRC6U1ycsOSwq0AyzAj8fMSm6q4HhQ","user_id":"@the-bridge-user:hs1","age":1172},{"type":"m.room.member","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@the-bridge-user:hs1","content":{"membership":"join"},"state_key":"@the-bridge-user:hs1","origin_server_ts":1630527559975,"unsigned":{"age":1206},"event_id":"$Y3XQmsUTwHPz68O9whnYVObrbQkJJuHd3jcRwpWbles","user_id":"@the-bridge-user:hs1","age":1206},{"type":"m.room.create","room_id":"!pFzlAcdlcfUYBlwrgD:hs1","sender":"@the-bridge-user:hs1","content":{"room_version":"org.matrix.msc2716","creator":"@the-bridge-user:hs1"},"state_key":"","origin_server_ts":1630527559920,"unsigned":{"age":1261},"event_id":"$uCV9s7uW1xlENfEBPcZ2hivKsqexyGKS3XDBUcOb84k","user_id":"@the-bridge-user:hs1","age":1261}],"start":"s15_0_0_0_0_0_0_0_0","end":"t1-1_0_0_0_0_0_0_0_0"}
2021/09/01 15:19:21 ============================================

2021/09/01 15:19:21 23133d0dbb53beb67b9d22502e16679ceca3dee79b6ded980cbeafba7730900e : Server logs:
Signature ok
subject=CN = hs2
[...]


2021-09-01 20:19:21,462 - synapse.handlers.federation_event - 746 - WARNING - GET-4-$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY - Error attempting to resolve state at missing prev_events
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/synapse/handlers/federation_event.py", line 711, in _resolve_state_at_missing_prevs
    remote_state = await self._get_state_after_missing_prev_event(
  File "/usr/local/lib/python3.8/site-packages/synapse/handlers/federation_event.py", line 777, in _get_state_after_missing_prev_event
    ) = await self.federation_client.get_room_state_ids(
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 370, in get_room_state_ids
    result = await self.transport_layer.get_room_state_ids(
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/transport/client.py", line 71, in get_room_state_ids
    return await self.client.get_json(
  File "/usr/local/lib/python3.8/site-packages/synapse/http/matrixfederationclient.py", line 1001, in get_json
    response = await self._send_request_with_optional_trailing_slash(
  File "/usr/local/lib/python3.8/site-packages/synapse/http/matrixfederationclient.py", line 385, in _send_request_with_optional_trailing_slash
    response = await self._send_request(request, **send_request_args)
  File "/usr/local/lib/python3.8/site-packages/synapse/http/matrixfederationclient.py", line 631, in _send_request
    raise exc
synapse.api.errors.HttpResponseException: 404: Not Found
2021-09-01 20:19:21,464 - synapse.handlers.federation_event - 637 - WARNING - GET-4-$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY - Pulled event $hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY failed history check.
2021-09-01 20:19:21,465 - synapse.handlers.federation_event - 605 - INFO - GET-4-$lonJkljNkygvTMCeOqLpS4YLwrKch1cyPLirnrjhmqw - Processing pulled event <FrozenEventV3 event_id='$lonJkljNkygvTMCeOqLpS4YLwrKch1cyPLirnrjhmqw', type='m.room.message', state_key=None>
2021-09-01 20:19:21,472 - synapse.handlers.federation_event - 605 - INFO - GET-4-$kqz1sVD1G5kyK6tZ5_UR6uLNb6dKZihbXHYi1yScmDc - Processing pulled event <FrozenEventV3 event_id='$kqz1sVD1G5kyK6tZ5_UR6uLNb6dKZihbXHYi1yScmDc', type='org.matrix.msc2716.insertion', state_key=None>
2021-09-01 20:19:21,479 - synapse.handlers.federation_event - 605 - INFO - GET-4-$SyQzaYC7H4vlGUrQ0YdmccOBkCg71sL0KnMn425oj1U - Processing pulled event <FrozenEventV3 event_id='$SyQzaYC7H4vlGUrQ0YdmccOBkCg71sL0KnMn425oj1U', type='org.matrix.msc2716.chunk', state_key=None>
2021-09-01 20:19:21,481 - synapse.handlers.federation_event - 1567 - INFO - GET-4-$SyQzaYC7H4vlGUrQ0YdmccOBkCg71sL0KnMn425oj1U - auth_events refers to events which are not in our calculated auth chain: {'$Y3XQmsUTwHPz68O9whnYVObrbQkJJuHd3jcRwpWbles'}
2021-09-01 20:19:21,481 - synapse.state - 461 - INFO - GET-4-$SyQzaYC7H4vlGUrQ0YdmccOBkCg71sL0KnMn425oj1U - Resolving state for !pFzlAcdlcfUYBlwrgD:hs1 with 2 groups
2021-09-01 20:19:21,482 - synapse.handlers.federation_event - 1606 - INFO - GET-4-$SyQzaYC7H4vlGUrQ0YdmccOBkCg71sL0KnMn425oj1U - After state res: updating auth_events with new state {('m.room.member', '@the-bridge-user:hs1'): '$Y3XQmsUTwHPz68O9whnYVObrbQkJJuHd3jcRwpWbles'}
2021-09-01 20:19:21,489 - synapse.handlers.federation_event - 605 - INFO - GET-4-$SrhE54fDdLcQI6dyxFE-QaTfSZhGEB0TkjUagSQ5AKc - Processing pulled event <FrozenEventV3 event_id='$SrhE54fDdLcQI6dyxFE-QaTfSZhGEB0TkjUagSQ5AKc', type='m.room.message', state_key=None>
2021-09-01 20:19:21,496 - synapse.handlers.federation_event - 605 - INFO - GET-4-$CSvwLw2nHOW0O3MkI2EoTOTPqNFxI5Zdn73tzdKayBs - Processing pulled event <FrozenEventV3 event_id='$CSvwLw2nHOW0O3MkI2EoTOTPqNFxI5Zdn73tzdKayBs', type='m.room.message', state_key=None>
2021-09-01 20:19:21,501 - synapse.handlers.federation_event - 605 - INFO - GET-4-$VtcB-yfsAMsd34RofKyrV9vqKLHRmuQOIlBDU3x1IDA - Processing pulled event <FrozenEventV3 event_id='$VtcB-yfsAMsd34RofKyrV9vqKLHRmuQOIlBDU3x1IDA', type='m.room.message', state_key=None>
2021-09-01 20:19:21,531 - synapse.access.http.8008 - 389 - INFO - GET-4 - ::ffff:192.168.112.1 - 8008 - {@charlie:hs2} Processed request: 0.275sec/-0.000sec (0.041sec, 0.009sec) (0.219sec/0.045sec/91) 5474B 200 "GET /_matrix/client/r0/rooms/%21pFzlAcdlcfUYBlwrgD:hs1/messages?dir=b&limit=100 HTTP/1.1" "Go-http-client/1.1" [16 dbevts]
2021-09-01 20:19:21,548 - synapse.http.matrixfederationclient - 287 - INFO - GET-5 - {GET-O-10} [hs1] Completed request: 200 OK in 0.01 secs, got 774 bytes - GET matrix://hs1/_matrix/federation/v1/backfill/%21pFzlAcdlcfUYBlwrgD%3Ahs1?v=%24hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY&limit=100
2021-09-01 20:19:21,550 - synapse.handlers.federation_event - 605 - INFO - GET-5-$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY - Processing pulled event <FrozenEventV3 event_id='$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY', type='org.matrix.msc2716.insertion', state_key=None>
2021-09-01 20:19:21,550 - synapse.handlers.federation_event - 622 - INFO - GET-5-$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY - De-outliering event $hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY
2021-09-01 20:19:21,552 - synapse.handlers.federation_event - 683 - INFO - GET-5-$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY - Event $hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY is missing prev_events ['$TiICyIaUJplRPSUJfVCHnjmtOmYtRFiwoWWnYvLIABg']: calculating state for a backwards extremity
2021-09-01 20:19:21,552 - synapse.handlers.federation_event - 705 - INFO - GET-5-$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY - Requesting state after missing prev_event $TiICyIaUJplRPSUJfVCHnjmtOmYtRFiwoWWnYvLIABg
2021-09-01 20:19:21,559 - synapse.http.matrixfederationclient - 592 - INFO - GET-5-$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY-$TiICyIaUJplRPSUJfVCHnjmtOmYtRFiwoWWnYvLIABg - {GET-O-11} [hs1] Got response headers: 404 Not Found
2021-09-01 20:19:21,559 - synapse.http.matrixfederationclient - 670 - WARNING - GET-5-$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY-$TiICyIaUJplRPSUJfVCHnjmtOmYtRFiwoWWnYvLIABg - {GET-O-11} [hs1] Request failed: GET matrix://hs1/_matrix/federation/v1/state_ids/%21pFzlAcdlcfUYBlwrgD%3Ahs1?event_id=%24TiICyIaUJplRPSUJfVCHnjmtOmYtRFiwoWWnYvLIABg: HttpResponseException('404: Not Found')
2021-09-01 20:19:21,560 - synapse.handlers.federation_event - 746 - WARNING - GET-5-$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY - Error attempting to resolve state at missing prev_events
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/synapse/handlers/federation_event.py", line 711, in _resolve_state_at_missing_prevs
    remote_state = await self._get_state_after_missing_prev_event(
  File "/usr/local/lib/python3.8/site-packages/synapse/handlers/federation_event.py", line 777, in _get_state_after_missing_prev_event
    ) = await self.federation_client.get_room_state_ids(
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/federation_client.py", line 370, in get_room_state_ids
    result = await self.transport_layer.get_room_state_ids(
  File "/usr/local/lib/python3.8/site-packages/synapse/federation/transport/client.py", line 71, in get_room_state_ids
    return await self.client.get_json(
  File "/usr/local/lib/python3.8/site-packages/synapse/http/matrixfederationclient.py", line 1001, in get_json
    response = await self._send_request_with_optional_trailing_slash(
  File "/usr/local/lib/python3.8/site-packages/synapse/http/matrixfederationclient.py", line 385, in _send_request_with_optional_trailing_slash
    response = await self._send_request(request, **send_request_args)
  File "/usr/local/lib/python3.8/site-packages/synapse/http/matrixfederationclient.py", line 631, in _send_request
    raise exc
synapse.api.errors.HttpResponseException: 404: Not Found
2021-09-01 20:19:21,560 - synapse.handlers.federation_event - 637 - WARNING - GET-5-$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY - Pulled event $hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY failed history check.
2021-09-01 20:19:21,567 - synapse.access.http.8008 - 389 - INFO - GET-5 - ::ffff:192.168.112.1 - 8008 - {@charlie:hs2} Processed request: 0.033sec/-0.000sec (0.009sec, 0.000sec) (0.003sec/0.001sec/7) 5474B 200 "GET /_matrix/client/r0/rooms/%21pFzlAcdlcfUYBlwrgD:hs1/messages?dir=b&limit=100 HTTP/1.1" "Go-http-client/1.1" [0 dbevts]
2021/09/01 15:19:21 ============== 23133d0dbb53beb67b9d22502e16679ceca3dee79b6ded980cbeafba7730900e : END LOGS ==============
```

Reply:
It seems like our [fake `prev_event`](https://github.com/matrix-org/synapse/blob/c586d6803a2adebcdd486be2d9eac1f62fd7d4ab/synapse/rest/client/room_batch.py#L241) is coming back to bite us 😣 and is causing any event that references the fake event to be rejected. We originally discussed this at https://github.com/matrix-org/matrix-doc/pull/2716#discussion_r598896168 (look for the `floating` discussion) to make the state float off on their own pretending it was in the depths of history somewhere we haven't fetched yet.

 - `$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY` is the insertion event for the chunk
 - `$TiICyIaUJplRPSUJfVCHnjmtOmYtRFiwoWWnYvLIABg` is the fake event
 
```
Event $hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY is missing prev_events ['$TiICyIaUJplRPSUJfVCHnjmtOmYtRFiwoWWnYvLIABg']: calculating state for a backwards extremity
Requesting state after missing prev_event $TiICyIaUJplRPSUJfVCHnjmtOmYtRFiwoWWnYvLIABg
WARNING - GET-5-$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY-$TiICyIaUJplRPSUJfVCHnjmtOmYtRFiwoWWnYvLIABg - {GET-O-14} [hs1] Request failed: GET matrix://hs1/_matrix/federation/v1/state_ids/%21yOGzjYcqVcKjTRlAxz%3Ahs1?event_id=%24tJUTwqLgFBCygrtCiMInLChmRCqNFdnbHdqaEbCaVnF: HttpResponseException('404: Not Found')
WARNING - GET-5-$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY - Error attempting to resolve state at missing prev_events
FederationError=ERROR 403: We can't get valid state history. event_id=$hbHXgulNKlmg9mkGQiBH5SSgq2NL4GO70sIIfHCa-YY
```

Reply:
It looks like the culprit is in https://github.com/matrix-org/synapse/pull/10645 which changed how `backfill` gets state and auths events.

If I restore `_get_state_for_room` and the old implementation in `backfill`, the Complement tests pass again.

Reply:
hmmm, maaayyybe we don't need to "[cause] the HS to ask for the state at the start of the chunk later." using `fake_prev_event_id` with the insertion events. The tests still pass if I revert https://github.com/matrix-org/synapse/commit/a8c53114ef74c474eb5bfd08aa92ed9d34977ca6.

And to jog the memory, here is the DAG in either scenario:

Connected to main DAG `prev_event` | With `fake_prev_event_id`
--- | ---
![](https://user-images.githubusercontent.com/558581/126261578-a6f6c3ec-bf5c-4c82-a128-66ba4a9ac079.png) | ![](https://user-images.githubusercontent.com/558581/127040602-e95ac36a-5e64-4176-904d-6abae2c95ae9.png)


@erikjohnston I could use some insight here ^ in case you have opinions on keeping the fake event or where/what we need to add to accommodate it if we need to use it still.

Otherwise, I will be digging into the past and new `backfill` implementation finding the key difference.

Reply:
> @erikjohnston I could use some insight here ^ in case you have opinions on keeping the fake event or where/what we need to add to accommodate it if we need to use it still.

If we don't have the fake event then the state of the chunks will be inherited from `A`, which will likely be wrong, so I think we still need the prev event.

---

It looks like that PR changed the way we handle fetching state for backfills, before we'd just get the state at the last event, whereas now we try and get the state for each prev event (including the fake event), which obviously fails. I'll talk to @richvdh to try and see what the correct thing to do here is, but my hunch is we probably *do* want to be able to handle the case where no one has the state for an event.

Reply:
I chatted with @richvdh briefly, and the conclusion was that a) that code is protecting against an attack where a remote server could arbitrarily replace the servers view of the current state of the room, but b) the mitigation we use is a sledgehammer and we should replace it with something more intelligent that both mitigates the attack and gracefully handles missing events.

I'm not sure what that looks like yet, but will have a think.

Reply:
Thanks for the checking this out! Since the Complement test also fails on `develop`, I think we're good to merge this and move forward.

I'll create a new separate issue to track this new unobtanium fake hammer we're going to forge 💪

Reply:
Tracked in https://github.com/matrix-org/synapse/issues/10764